File: src/game/level/ItemChart.java
package game.level;
import game.Vector;
import game.item.Item;
import game.item.ItemType;

import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

import javax.imageio.stream.FileImageInputStream;


public class ItemChart {
	public List<SpawnPair> spawns = new ArrayList<>();
	
	public static class SpawnPair {
		Vector pos;
		ItemType itype;
		
		public SpawnPair() {}
		
		public SpawnPair(Vector pos, ItemType itype) {
			this.pos = pos;
			this.itype = itype;
		}
		
		public Item instantiate() {
			return itype.instantiate(pos);
		}
	}
	
	public List<Item> instantiate() {
		List<Item> items = new ArrayList<>();
		for (SpawnPair pair : spawns) {
			items.add(pair.instantiate());
		}
		return items;
	}
	
	public void saveToFile(String filename) {
		PrintWriter outputWriter = null;
		try {
			outputWriter = new PrintWriter(filename);
			outputWriter.println(spawns.size());
			for (SpawnPair pair : spawns) {
				outputWriter.println(pair.pos.x + " " + pair.pos.y + " " + pair.itype.getCode());
			}
			outputWriter.flush();
			System.out.println("Saved item chart to '" + filename + "'.");
		} catch (IOException e) {
			e.printStackTrace();
		} finally {
			outputWriter.close();
		}
	}
	
	public static ItemChart loadFromFile(String filename) throws IOException {
		FileImageInputStream reader = null;
		ItemChart chart = new ItemChart();
		try {
			reader = new FileImageInputStream(new File(filename));
			
			Scanner s = new Scanner(reader.readLine());
			int numSpawns = s.nextInt();
			s.close();
			
			for (int i = 0; i < numSpawns; i++) {
				SpawnPair pair = new SpawnPair();
				s = new Scanner(reader.readLine());
				pair.pos = new Vector(s.nextFloat(), s.nextFloat());
				pair.itype = ItemType.fromCode(s.nextInt());
				s.close();
				chart.spawns.add(pair);
			}
			System.out.println("Loaded item chart from " + filename);
		} catch (IOException e) {
			throw new IOException("Failed to load item chart: " + filename);
		} finally {
			try {
				reader.close();
			} catch (Exception e) {}
		}
		return chart;
	}
}

File: ts/game/level/ItemChart.ts
import { Vector } from '../Vector.js';
import { Item } from '../item/Item.js';
import { ItemType, instantiateItem, fromCode } from '../item/ItemType.js';

export class ItemChart {
  spawns: SpawnPair[] = [];
}

export class SpawnPair {
  constructor(public pos: Vector, public itype: ItemType) {}
  instantiate(): Item {
    return instantiateItem(this.itype, this.pos);
  }
}

export function instantiateItemChart(chart: ItemChart): Item[] {
  return chart.spawns.map((s) => s.instantiate());
}

export function saveItemChart(_chart: ItemChart, _filename: string) {
  console.warn('saveItemChart is not supported in the browser');
}

export async function loadItemChart(filename: string): Promise<ItemChart> {
  const res = await fetch(filename);
  const data = (await res.text()).trim().split(/\r?\n/);
  const chart = new ItemChart();
  const num = parseInt(data[0], 10);
  for (let i = 1; i <= num; i++) {
    const [x, y, code] = data[i].split(' ').map(Number);
    chart.spawns.push(new SpawnPair(new Vector(x, y), fromCode(code)));
  }
  return chart;
}
